pkg/ccl/backupccl/restore_planning.go

```diff
@@ -158,6 +159,251 @@ func synthesizePGTempSchema(
+	if err := bumpDescIDSequenceKey(ctx, p, databasesByID, schemasByID, tablesByID, typesByID); err != nil {
+		return err
+	}
+
+	for _, table := range systemTablesWithConflictingIDs {
+		log.Infof(ctx, "moving system table %s to non-conflicting ID", table.GetName())
+		if err := moveConflictingSystemTable(ctx, p, table); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// moveConflictingSystemTable moves a system table to a higher ID.
+//
+// We do the move by creating new tables in the system database. Data from the
+// conflicting tables is moved to the temporary tables. The namespace and
+// descriptor entry for the conflicting table is removed, and then the new table
+// is renamed.
+//
+// TODO(ssd): We currently do all of the descriptor modification "bare handed"
+// by writing to KV directly. Should we be using WriteDescToBatch and friends
+// instead? I avoided it for now because I was having some trouble with
+// validation when using it.
+func moveConflictingSystemTable(
+	ctx context.Context, p sql.PlanHookState, oldTable catalog.TableDescriptor,
+) error {
+	ctx, span := tracing.ChildSpan(ctx, "backupccl.moveConflictingSystemTable")
+	defer span.Finish()
+
+	codec := p.ExecCfg().Codec
+	return sql.DescsTxn(ctx, p.ExecCfg(), func(ctx context.Context, txn *kv.Txn, descsCol *descs.Collection) error {
+		b := txn.NewBatch()
+		// Create a new table with the same schema as the table we want to
+		// move.
+		newTable := oldTable.NewBuilder().BuildCreatedMutable().(*tabledesc.Mutable)
+
+		// Allow the upgrades package to dynamically assign an ID to the new system
+		// table.
+		newTable.ID = descpb.InvalidID
+		newTable.Name = fmt.Sprintf("crdb_internal_copy_%s", newTable.Name)
+		id, created, err := upgrades.CreateSystemTableInTxn(ctx, p.ExecCfg().DB, txn, codec, newTable)
+		if err != nil {
+			return err
+		}
+		if !created {
+			return errors.New("temporary table for system table already exists")
+		}
+
+		// Copy data from the old table to the new table.
+		insertStmt := fmt.Sprintf("INSERT INTO system.%s (SELECT * FROM system.%s)", newTable.Name, oldTable.GetName())
+		if _, err := p.ExecCfg().InternalExecutor.Exec(ctx, "copy-to-temp-table", txn, insertStmt); err != nil {
+			return err
+		}
+
+		// Delete the old namespace entry for the new table.
+		b.Del(catalogkeys.EncodeNameKey(codec, newTable))
+
+		// Delete the descriptor entry for the old system table.
+		b.Del(catalogkeys.MakeDescMetadataKey(codec, oldTable.GetID()))
+		descsCol.AddDeletedDescriptor(oldTable.GetID())
+
+		// Overwrite the namespace entry for the system table with our new id, and update the
+		// descriptor of the new table with the old table's name.
+		newTable.ID = id
+		newTable.Name = oldTable.GetName()
+		if err := descsCol.WriteDescToBatch(ctx, false /* kvTrace */, newTable, b); err != nil {
+			return err
+		}
+		b.Put(catalogkeys.EncodeNameKey(codec, newTable), newTable.ID)
+
+		if err := txn.Run(ctx, b); err != nil {
+			return err
+		}
```

Comment:
So I'm generating fixtures for older backups in 21.1 onwards and I'm getting a validation error which might be related to your comment about having to b.Put the new descriptor instead of going through WriteDescToBatch. This got me trying an alternate approach of:

- CREATE TABLE copy AS SELECT * from old_system;
- DROP TABLE old_system;
- ALTER TABLE copy RENAME TO old_system;

instead of this KV level manipulation. This does mean that we have a dropped descriptor waiting out GC, but is this a big deal? We already drop defaultdb and postgres as a part of restore anyways.

Reply:
@adityamaru I like the idea of doing it all via SQL.

Reply:
> This does mean that we have a dropped descriptor waiting out GC, but is this a big deal?

Will we also have a GCJob waiting around that is going to clear out the table keyspace?  Not a blocker, just something to double check when we take a look.

Reply:
Can we special case the logic that runs here to behave differently for descriptors in the system database?

Reply:
Good idea, that's what I ended up doing.